onnxscript/values.py

```diff
@@ -183,77 +183,73 @@
 
         return fun
 
+    def wrap(self, inputs):
+        """Adapts inputs into representation used by onnxscript eager mode.
+
+        This does the following transformations:
+        * It adds an onnxscript Tensor wrapper around numpy arrays, which
+        allows the use of overloaded operators like + to be controlled by onnxscript.
+        * It also provides a promotion of scalars into tensors as a convenience.
+        This is needed to complement the similar promotion supported by the
+        onnxscript converter (for example, when an attribute is promoted and used
+        as an input argument).
+
+        Args:
+            inputs: a list/tuple of inputs to an ONNX function
+
+        Returns:
+            a pair (wrapped_inputs, flag) where flag indicates whether any numpy array
+            was wrapped into a Tensor.
+        """
+        has_array = False
+
+        def do_wrap(input):
+            if isinstance(input, np.ndarray):
+                nonlocal has_array
+                has_array = True
+                return tensor.Tensor(input)
+            elif isinstance(input, tensor.Tensor):
+                return input
+            elif isinstance(input, (bool, int, float)):
+                return tensor.Tensor(np.array(input))
+            elif input is None:
+                return None
+            elif isinstance(input, list):
+                return [do_wrap for elt in input]
+            elif isinstance(input, tuple):
+                return tuple([do_wrap(elt) for elt in input])
```

Comment:
We should probably disable this rule in these lines as a generator can have a bigger overhead for this type of uses.

Reply:
Will it? I don't fully understand what the compiler does in the two cases. Does list-comprehension work-in-place without using a generator? Why can't `tuple(e for x in l)` produce similar code to `[e for x in l]`?

Reply:
`tuple(e for x in l)`:

a generator is first created by `e for x in l`, then `tuple` consumes it to initialize the tuple. 

`tuple([e for x in l])`:

a list is created by the comprehension, then used by `tuple`.

Last time I tested the generator version was slower. This may have changed, and the generator version is admittedly cleaner.